### ISSUE TITLE
Request for modification events of label layers

### ISSUE THREAD

**Issue by a-ludi** [napari/napari#4194]
Currently, the only way of tracking changes of a label layer is to connect to the `set_data` event which is very general. In our plugin, we would like to track changes in a more detailed way in order to keep track of labels without analyzing all the label data.

For the moment, we are connecting to `set_data` and parse the history to see what has changed. This is complicated, error-prone and fragile. Furthermore, accessing the history attributes has been officially deprecated.

In general, it would be a much better solution to add a designated event which gets fired when the layer data changes due to one of the built-in tools (eraser/brush, fill). It would require minimal changes in [the `paint` method](https://github.com/napari/napari/blob/d5d25cdb195601264053e38da4ecadaa791c0ba2/napari/layers/labels/labels.py#L1295) and [the `fill` method](https://github.com/napari/napari/blob/d5d25cdb195601264053e38da4ecadaa791c0ba2/napari/layers/labels/labels.py#L1203).

I would suggest to add an event named e.g. `data_diff` and pass the same object that is recorded in the history.

**Comment by jni** [comment]
Yup, I've needed this before @a-ludi so thanks for the issue, hoping we can fix soon, or if you would like to make a pull request (contributor guide [here](https://napari.org/developers/contributing.html)), I would be very happy to help you get it in.

**Comment by Czaki** [comment]
The problem with history is that it store only 2D changes. So if you do 3D painting or 3D fill you will not get real changes.

**Comment by jni** [comment]
@Czaki that's no longer true (or if it is it's a bug). I changed the history to store the exact indices and values that are written to the array.

**Comment by Czaki** [comment]
> @Czaki that's no longer true (or if it is it's a bug). I changed the history to store the exact indices and values that are written to the array.

It is hard to parse this code from mobile, but if I good understand this code then adding events may introduce performance problem. And if you throttle events to be emitted after edit is done then you may lost information.

**Comment by jni** [comment]
I think if we only emit on click release (ie one history item), then I don't expect the performance penalty to be huge. But, we will see.

**Comment by a-ludi** [comment]
First of all, thanks for all the feedback.

Regarding the performance, I do not see a big problem here unless somebody does a lot of work in the handler. I mean, in my current implementation I am actually parsing part of the history on every `set_data` event which is certainly more expensive than the debated solution.

I have been thinking about my proposal and came to the conclusion that it is incomplete because it only introduces events for adding to the history. In fact, one needs to observe all changes to the history state (adding, undo and redo) to keep track of things.

Therefore, I think the events need to be located in `_save_history` and `_load_history`. I will try to create a pull request one of these days.

**Comment by Czaki** [comment]
> I do not see a big problem here unless somebody does a lot of work in the handler.

We cannot assume this :(. But aggregated history suggested by @jni should be the best solution.

**Comment by a-ludi** [comment]
I do not know how to best implement the "aggregated history" event, so somebody else must do that. It would definitely be a great option.

But I really do not see the problem with the "immediate history" event. I believe it is the responsibility of the plugin author to maintain performance. Plugin authors must be careful with event handlers in any kind of event really. Right now we are using the `set_data` event which is fired even more often.
